tbdex/rfq/rfq.go

```diff
@@ -166,6 +173,213 @@ func (r *RFQ) verifyPrivateData() error {
+		}
+	}
+
+	// Verify payin method
+	err = r.verifyPaymentMethod(
+		r.Data.Payin.Kind,
+		r.Data.Payin.PaymentDetailsHash,
+		r.PrivateData,
+		convertPayinMethods(offering.Data.Payin.Methods),
+		"payin",
+	)
+	if err != nil {
+		return fmt.Errorf("failed to verify payin method: %w", err)
+	}
+
+	// Verify payout method
+	err = r.verifyPaymentMethod(
+		r.Data.Payout.Kind,
+		r.Data.Payout.PaymentDetailsHash,
+		r.PrivateData,
+		convertPayoutMethods(offering.Data.Payout.Methods),
+		"payout",
+	)
+
+	if err != nil {
+		return fmt.Errorf("failed to verify payout method: %w", err)
+	}
+
+	err = r.verifyClaims(offering.Data.RequiredClaims)
+	if err != nil {
+		return fmt.Errorf("failed to verify claims: %w", err)
+	}
+
+	return nil
+
+}
+
+func convertPayinMethods(payinMethods []offering.PayinMethod) []offering.PaymentMethod {
+	paymentMethods := make([]offering.PaymentMethod, len(payinMethods))
+	for i, method := range payinMethods {
+		paymentMethods[i] = method
+	}
+	return paymentMethods
+}
+
+func convertPayoutMethods(payoutMethods []offering.PayoutMethod) []offering.PaymentMethod {
+	paymentMethods := make([]offering.PaymentMethod, len(payoutMethods))
+	for i, method := range payoutMethods {
+		paymentMethods[i] = method
+	}
+	return paymentMethods
+}
+
+func (r *RFQ) verifyPaymentMethod(
+	selectedPaymentKind string,
+	selectedPaymentDetailsHash string,
+	privateData *PrivateData,
+	offeringPaymentMethods []offering.PaymentMethod,
+	payDirection PayDirection,
+) error {
+	// Filter allowed payment methods by selectedPaymentKind
+	var allowedPaymentMethods []offering.PaymentMethod
+	for _, paymentMethod := range offeringPaymentMethods {
+		if paymentMethod.GetKind() == selectedPaymentKind {
+			allowedPaymentMethods = append(allowedPaymentMethods, paymentMethod)
+		}
+	}
+
+	// If no matching payment methods are found, throw an error
+	if len(allowedPaymentMethods) == 0 {
+		var allowedPaymentMethodKinds []string
+		for _, paymentMethod := range offeringPaymentMethods {
+			allowedPaymentMethodKinds = append(allowedPaymentMethodKinds, paymentMethod.GetKind())
+		}
+		return fmt.Errorf(
+			"offering does not support rfq's %sMethod kind. (rfq) %s was not found in: [%s] (offering)",
+			payDirection, selectedPaymentKind, strings.Join(allowedPaymentMethodKinds, ", "),
+		)
+	}
+
+	for _, paymentMethod := range allowedPaymentMethods {
+		if paymentMethod.GetRequiredPaymentDetails() == nil {
+			// If requiredPaymentDetails is omitted, and selectedPaymentDetailsHash is also omitted, we have a match
+			if selectedPaymentDetailsHash == "" {
+				return nil
+			}
+			// selectedPaymentDetailsHash is present even though requiredPaymentDetails is omitted.
+			return errors.New("rfq paymentDetails must be omitted when offering requiredPaymentDetails is omitted")
+		}
+
+		// todo do we actually want to return error in this case where requiredPaymentDetails is present but privateData is nil?
```

Comment:
question for @mistermoe 

in the context of `rfq.VerifyOfferingRequirements()` which gets called some time after receiving RFQ via `POST /exchanges/:id`, do want to return an error where requiredPaymentDetails is present but privateData is nil?